File: palettier-core/src/main/java/be/asmolabs/palettier/core/color/Rgb.java
package be.asmolabs.palettier.core.color;

/**
 * Couleur sRGB, chaque composante exprimee dans l'intervalle [0, 1].
 */
public record Rgb(double r, double g, double b) {

    public Rgb {
        r = clamp01(r);
        g = clamp01(g);
        b = clamp01(b);
    }

    public static Rgb ofHex(String hex) {
        String value = hex.startsWith("#") ? hex.substring(1) : hex;
        if (value.length() == 3) {
            value = "" + value.charAt(0) + value.charAt(0)
                    + value.charAt(1) + value.charAt(1)
                    + value.charAt(2) + value.charAt(2);
        }
        if (value.length() != 6) {
            throw new IllegalArgumentException("Code hexadecimal invalide : " + hex);
        }
        int packed = Integer.parseInt(value, 16);
        return new Rgb(((packed >> 16) & 0xFF) / 255.0,
                ((packed >> 8) & 0xFF) / 255.0,
                (packed & 0xFF) / 255.0);
    }

    public String toHex() {
        return "#%02X%02X%02X".formatted(to255(r), to255(g), to255(b));
    }

    /** Luminance relative (WCAG), utile pour choisir une couleur de texte lisible sur la pastille. */
    public double relativeLuminance() {
        return 0.2126 * Colors.srgbToLinear(r)
                + 0.7152 * Colors.srgbToLinear(g)
                + 0.0722 * Colors.srgbToLinear(b);
    }

    private static int to255(double v) {
        return (int) Math.round(v * 255.0);
    }

    private static double clamp01(double v) {
        return Math.clamp(v, 0.0, 1.0);
    }
}

File: palettier-kmp/core-domain/src/commonMain/kotlin/be/asmolabs/palettier/domain/color/Rgb.kt
package be.asmolabs.palettier.domain.color

import kotlin.math.roundToInt

/**
 * Couleur sRGB, chaque composante exprimee dans l'intervalle [0, 1].
 *
 * <p>Les composantes sont tronquees a l'intervalle plutot que refusees, exactement comme
 * le constructeur compact cote Java. Ce n'est pas de la complaisance : les conversions
 * lumiere lineaire / sRGB rendent des valeurs qui debordent d'un milliardieme, et lever
 * une exception la-dessus casserait des calculs parfaitement justes.</p>
 */
class Rgb private constructor(val r: Double, val g: Double, val b: Double) {

    fun toHex(): String = "#" + listOf(r, g, b).joinToString("") { channel ->
        (channel * 255.0).roundToInt().toString(16).uppercase().padStart(2, '0')
    }

    /** Luminance relative (WCAG), utile pour choisir une couleur de texte lisible sur la pastille. */
    fun relativeLuminance(): Double =
        0.2126 * srgbToLinear(r) + 0.7152 * srgbToLinear(g) + 0.0722 * srgbToLinear(b)

    override fun equals(other: Any?) = other is Rgb && r == other.r && g == other.g && b == other.b
    override fun hashCode() = (r.hashCode() * 31 + g.hashCode()) * 31 + b.hashCode()
    override fun toString() = toHex()

    operator fun component1() = r
    operator fun component2() = g
    operator fun component3() = b

    companion object {
        /** Seule voie de construction : elle tronque, comme le constructeur compact Java. */
        operator fun invoke(r: Double, g: Double, b: Double) =
            Rgb(r.coerceIn(0.0, 1.0), g.coerceIn(0.0, 1.0), b.coerceIn(0.0, 1.0))

        fun ofHex(hex: String): Rgb {
            val raw = hex.removePrefix("#")
            val value = when (raw.length) {
                3 -> raw.flatMap { listOf(it, it) }.joinToString("")
                6 -> raw
                else -> throw IllegalArgumentException("Code hexadecimal invalide : $hex")
            }
            val packed = value.toIntOrNull(16)
                ?: throw IllegalArgumentException("Code hexadecimal invalide : $hex")
            return Rgb(
                ((packed shr 16) and 0xFF) / 255.0,
                ((packed shr 8) and 0xFF) / 255.0,
                (packed and 0xFF) / 255.0,
            )
        }
    }
}

/** Couleur dans l'espace CIE L*a*b* (illuminant D65, observateur 2 degres). */
data class Lab(val l: Double, val a: Double, val b: Double)
